### ISSUE TITLE
Keras saved model returns different result than original model using Batch Normalization on multiple GPUs (Distributed training)

### ISSUE THREAD

**Issue by Lucasbravo1** [tensorflow/tensorflow#46488]
**System information**
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow): No
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): Ubuntu 18.04
- Mobile device (e.g. iPhone 8, Pixel 2, Samsung Galaxy) if the issue happens on mobile device:
- TensorFlow installed from (source or binary): Binary via pip3
- TensorFlow version (use command below): 2.3.1
- Python version: 3.6.9
- Bazel version (if compiling from source):
- GCC/Compiler version (if compiling from source):
- CUDA/cuDNN version: 10.1
- GPU model and memory: Geforce GTX 1080 Ti 4x11GB

I got a different result when using evaluate function on a saved model when compare with the original model. This only happens when Batch Normalization is included in the model and when training on multiple GPUs with MirroredStrategy.

Here is my model
```
with strategy.scope():
  model = tf.keras.Sequential([
      tf.keras.layers.Conv2D(32, 3, activation='relu', input_shape=(28, 28, 1)),
      tf.keras.layers.MaxPooling2D(),
      tf.keras.layers.BatchNormalization(),
      tf.keras.layers.Flatten(),
      tf.keras.layers.Dense(64, activation='relu'),
      tf.keras.layers.Dense(10)
  ])

  model.compile(loss=tf.keras.losses.SparseCategoricalCrossentropy(from_logits=True),
                optimizer=tf.keras.optimizers.Adam(),
                metrics=['accuracy'])
```

Multiple GPUs with MirroredStrategy
```
strategy = tf.distribute.MirroredStrategy()
print('Number of devices: {}'.format(strategy.num_replicas_in_sync))
```
Output
```
Number of devices: 2
```

Evaluate after training
```
eval_loss, eval_acc = model.evaluate(eval_dataset)

print('Eval loss: {}, Eval Accuracy: {}'.format(eval_loss, eval_acc))
```

Output
```
79/79 [==============================] - 0s 5ms/step - loss: 0.0424 - accuracy: 0.9884
Eval loss: 0.04239395260810852, Eval Accuracy: **0.9883999824523926**
```

Save model and evaluate again
```
path = 'saved_model/'
model.save(path, save_format='tf')
with strategy.scope():
  replicated_model = tf.keras.models.load_model(path)
  replicated_model.compile(loss=tf.keras.losses.SparseCategoricalCrossentropy(from_logits=True),
                           optimizer=tf.keras.optimizers.Adam(),
                           metrics=['accuracy'])

  eval_loss, eval_acc = replicated_model.evaluate(eval_dataset)
  print ('Eval loss: {}, Eval Accuracy: {}'.format(eval_loss, eval_acc))
```
Output
```
79/79 [==============================] - 0s 6ms/step - loss: 0.0424 - accuracy: 0.9883
Eval loss: 0.04239019751548767, Eval Accuracy: **0.9883000254631042**
```

 - Without BN

Output from evaluate
```
79/79 [==============================] - 0s 6ms/step - loss: 0.0450 - accuracy: 0.9837
Eval loss: 0.04498908668756485, Eval Accuracy: **0.9836999773979187**
```

Save model and repeat evaluate
```
79/79 [==============================] - 0s 4ms/step - loss: 0.0450 - accuracy: 0.9837
Eval loss: 0.04498908668756485, Eval Accuracy: **0.9836999773979187**
```

I have searched for similar issues and thought that this is because [BN computes differently during training and interference ](https://keras.io/api/layers/normalization_layers/batch_normalization/) but when I tried with one GPU, this issue didn't occur.

[Code for reproduce results](https://colab.research.google.com/drive/14iv88UJwFSv1SaVzmEl9RMOa1EH8IiRG?usp=sharing)

**Comment by nikitamaia** [comment]
Hi @Lucasbravo1, I've just run the code in TF2.4 with two GPUs and if I don't recompile the mode, I'm actually seeing the same accuracy for the BN model. Can you [run the code in this gist](https://colab.research.google.com/gist/nikitamaia/007bfe2a1213b9e9508c09be617b440f/github_issue_46488.ipynb#scrollTo=5Y04_YwwNnmY) on two GPUs and let me know if you see the same or if I'm missing something?
I don't think you need to recompile the model, see the[ note in the docs here](https://www.tensorflow.org/tutorials/distribute/save_and_load) "After restoring the model, you can continue training on it, even without needing to call compile() again, since it is already compiled before saving"

**Comment by Lucasbravo1** [comment]
Hi @nikitamaia, thank you very much for your support. 

Maybe I found where is the problem. If I run your code in TF2.3.1 the problem occurs but this works on TF2.4. [Here is the code on Colab](https://drive.google.com/file/d/139mdSKbeb-uSnf0hbLWYexuMARS5_Qzo/view?usp=sharing). Note that I reinstalled tensorflow from version 2.4.1 to 2.3.1. I haven't checked with multiple GPUs yet because my CUDA version is 11.1. I think tensorflow hasn't supported this version yet [link](https://www.tensorflow.org/install/source#gpu)?

**Comment by nikitamaia** [comment]
Yep, you can track support for 11.1 and 11.2 in #46093.

**Comment by Lucasbravo1** [comment]
Ok, thanks @nikitamaia. I will have a look. Hope this works with multiple gpus.

**Comment by nikitamaia** [comment]
I just ran some tests for the multi GPU case and there is a small difference in the loss values (accuracy is the same). 
After further investigation, it seems this is not totally unexpected. The model is saved as a non distributed model. Batch norm is a local statistic that happens on each replica, so when the model is reloaded the values are averaged across all of the replicas. You should not see this discrepancy if you use [SyncBatchNormalization](https://www.tensorflow.org/api_docs/python/tf/keras/layers/experimental/SyncBatchNormalization), although training might be slower.

**Comment by Lucasbravo1** [comment]
Thanks much @nikitamaia
